File: Napakalaki-Game(PDOO)/NapakalakiJava/src/napakalakiGame/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author joaquin
 */
public class SpecificBadConsequence extends BadConsequence{
    private ArrayList<TreasureKind> specificHiddenTreasures= new ArrayList();
    private ArrayList<TreasureKind> specificVisibleTreasures= new ArrayList();
    
    public SpecificBadConsequence(String text , int levels , ArrayList<TreasureKind> tVisible , ArrayList<TreasureKind> tHidden){
         super(text,levels);
         this.specificVisibleTreasures=tVisible;
         this.specificHiddenTreasures=tHidden;
    }
    
    public ArrayList<TreasureKind> getSpecificVisibleTreasures(){
        return this.specificVisibleTreasures;
    }
    
    public ArrayList<TreasureKind> getSpecificHiddenTreasures(){
        return this.specificHiddenTreasures;
    }
    @Override
    public boolean isEmpty(){
      boolean salida=false;
      if(specificVisibleTreasures.isEmpty() && specificHiddenTreasures.isEmpty())
        salida=true;
      return salida;
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t){
      if(specificVisibleTreasures.contains(t))
              specificVisibleTreasures.remove(t);
    }
    @Override
    public void substractHiddenTreasure(Treasure t){
      if(specificHiddenTreasures.contains(t))
              specificHiddenTreasures.remove(t);
     }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v ,ArrayList<Treasure> h ){
        BadConsequence salida = null;
        ArrayList<TreasureKind> copiaVisible = new ArrayList();
        ArrayList<TreasureKind> copiaHidden = new ArrayList();
        ArrayList<TreasureKind> specificVisible = new ArrayList(specificVisibleTreasures);
        ArrayList<TreasureKind> specificHidden = new ArrayList(specificHiddenTreasures);

        if(!specificVisibleTreasures.isEmpty()){
            boolean salir= false;
        for(Treasure pv:v){
            if(salir){
            specificVisible.remove(pv.getType());
            salir=false;
            }
            for(TreasureKind bv:specificVisible){
                if(pv.getType()==bv && !salir){
                    copiaVisible.add(bv);
                    salir=true;
                }
            }
        }}
        if(!specificHiddenTreasures.isEmpty()){
            boolean salir=false;
        for(Treasure ph:h){
            if(salir){
            specificHidden.remove(ph.getType());
            salir=false;
            }
            for(TreasureKind bh:specificHidden){
                if(ph.getType()==bh && !salir){
                    copiaHidden.add(bh);
                    salir=true;
                }
            }
        }}
        salida=new SpecificBadConsequence(super.getText(),super.getLevels(),copiaVisible,copiaHidden);
        
    return salida;
    }
    @Override
    public String toString(){
        return "\ntext = " + super.getText() + "\nlevels = " + Integer.toString(super.getLevels()) +  "\nSpecific-Visible-treasures = " + specificVisibleTreasures.toString()
                + "\nSpecific-hidden-treasures = " + specificHiddenTreasures.toString();
    
    }
}

File: Napakalaki-Game(PDOO)/NapakalakiRuby/lib/specific_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'card_dealer'
require 'bad_consequence.rb'
class SpecificBadConsequence < BadConsequence
  
  attr_reader:specificVisibleTreasures ,:specificHiddenTreasures
  public_class_method:new
  
  def initialize(aText,someLevels,someSpecificVisibleTreasures,someSpecificHiddenTreasures)
    super(aText,someLevels)
    @specificVisibleTreasures=someSpecificVisibleTreasures
    @specificHiddenTreasures= someSpecificHiddenTreasures
  end
  
  def isEmpty()
    salida=false
    if(@specificVisibleTreasures.empty? && @specificHiddenTreasures.empty?)
      salida=true
    end
    return salida
  end
  
   def substractVisibleTreasure(t)
    if(@specificVisibleTreasures.include?(t))
              @specificVisibleTreasures.remove(t)
    end
  end
  
  def substractHiddenTreasure(t)
    if(@specificHiddenTreasures.include?(t))
              @specificHiddenTreasures.remove(t)
    end
  end
  
  def adjustToFitTreasureLists(v,h)
        salida = nil
        copiaVisible = Array.new
        copiaHidden = Array.new
        specificVisible = Array.new(@specificVisibleTreasures)
        specificHidden = Array.new(@specificHiddenTreasures)
          if(!@specificVisibleTreasures.empty?)
            salir= false
            v.each do |pv|
            if(salir)
            specificVisible.delete(pv.getType())
            salir=false
            end
            specificVisible.each do |bv|
                if(pv.getType()==bv && !salir)
                    copiaVisible.push(bv)
                    salir=true
                end
            end
        end
      end
      if(!@specificHiddenTreasures.empty?)
            salir=false
            h.each do |ph|
            if(salir)
            specificHidden.delete(ph.getType())
            salir=false
            end
            specificHidden.each do |bh|
                if(ph.getType()==bh && !salir)
                    copiaHidden.push(bh)
                    salir=true
                end
            end
        end
        end
        salida=SpecificBadConsequence.new(text,levels,copiaVisible,copiaHidden)
    return salida
  end
  
  def to_s
    "texto: #{text} \n niveles: #{levels} \n nSpecificHiddenTreassures: #{@specificHiddenTreasures} \n nSpecificVisibleTreasures: #{@specificVisibleTreasures}"
  end
end
